src/auth/pluggable-auth-client.ts

```diff
@@ -0,0 +1,315 @@
+   *   the external account JSON credential file.
+   * @param additionalOptions Optional additional behavior customization
+   *   options. These currently customize expiration threshold time and
+   *   whether to retry on 401/403 API request errors.
+   */
+  constructor(
+    options: PluggableAuthClientOptions,
+    additionalOptions?: RefreshOptions
+  ) {
+    super(options, additionalOptions);
+    if (!options.credential_source.executable) {
+      throw new Error('No valid Pluggable Auth "credential_source" provided.');
+    }
+    this.command = options.credential_source.executable.command;
+    if (!this.command) {
+      throw new Error('No valid Pluggable Auth "credential_source" provided.');
+    }
+    // Check if the provided timeout exists and if it is valid.
+    if (options.credential_source.executable.timeout_millis === undefined) {
+      this.timeoutMillis = DEFAULT_EXECUTABLE_TIMEOUT_MILLIS;
+    } else {
+      this.timeoutMillis = options.credential_source.executable.timeout_millis;
+      if (
+        this.timeoutMillis < MINIMUM_EXECUTABLE_TIMEOUT_MILLIS ||
+        this.timeoutMillis > MAXIMUM_EXECUTABLE_TIMEOUT_MILLIS
+      ) {
+        throw new Error(
+          `Timeout must be between ${MINIMUM_EXECUTABLE_TIMEOUT_MILLIS} and ` +
+            `${MAXIMUM_EXECUTABLE_TIMEOUT_MILLIS} milliseconds.`
+        );
+      }
+    }
+
+    this.outputFile = options.credential_source.executable.output_file;
+
+    this.handler = new PluggableAuthHandler({
+      command: this.command,
+      timeoutMillis: this.timeoutMillis,
+      outputFile: this.outputFile,
+    });
+  }
+
+  /**
+   * Triggered when an external subject token is needed to be exchanged for a
+   * GCP access token via GCP STS endpoint.
+   * This uses the `options.credential_source` object to figure out how
+   * to retrieve the token using the current environment. In this case,
+   * this calls a user provided executable which returns the subject token.
+   * The logic is summarized as:
+   * 1. Validated that the executable is allowed to run. The
+   *    GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES environment must be set to
+   *    1 for security reasons.
+   * 2. If an output file is specified by the user, check the file location
+   *    for a response. If the file exists and contains a valid response,
+   *    return the subject token from the file.
+   * 3. Call the provided executable and return response.
+   * @return A promise that resolves with the external subject token.
+   */
+  async retrieveSubjectToken(): Promise<string> {
+    // Check if the executable is allowed to run.
+    if (process.env[GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES] !== '1') {
+      throw new Error(
+        'Pluggable Auth executables need to be explicitly allowed to run by ' +
+          'setting the GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES environment ' +
+          'Variable to 1.'
+      );
+    }
+
+    let executableResponse: ExecutableResponse | undefined;
```

Comment:
This should default to `= undefined`, otherwise this would produce an error at runtime if `#outputFile`  is falsy